lib/services/local/usagereporter.go

```diff
@@ -0,0 +1,544 @@
+		Help:      "a count of event batches that failed to submit",
+	})
+
+	usageEventsDropped = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: teleport.MetricNamespace,
+		Name:      teleport.MetricUsageEventsDropped,
+		Help:      "a count of events dropped due to submission buffer overflow",
+	})
+
+	usagePrometheusCollectors = []prometheus.Collector{
+		usageEventsSubmitted, usageBatchesTotal, usageEventsRequeuedTotal,
+		usageBatchSubmissionDuration, usageBatchesSubmitted, usageBatchesFailed,
+		usageEventsDropped,
+	}
+)
+
+// DiscardUsageReporter is a dummy usage reporter that drops all events.
+type DiscardUsageReporter struct{}
+
+func (d *DiscardUsageReporter) SubmitAnonymizedUsageEvents(event ...services.UsageAnonymizable) error {
+	// do nothing
+	return nil
+}
+
+// NewDiscardUsageReporter creates a new usage reporter that drops all events.
+func NewDiscardUsageReporter() *DiscardUsageReporter {
+	return &DiscardUsageReporter{}
+}
+
+// submitFunc is a func that submits a batch of usage events.
+type UsageSubmitFunc func(reporter *UsageReporter, batch []*prehogapi.SubmitEventRequest) error
+
+type UsageReporter struct {
+	// Entry is a log entry
+	*logrus.Entry
+
+	// clock is the clock used for the main batching goroutine
+	clock clockwork.Clock
+
+	// submitClock is the clock used for the submission goroutine
+	submitClock clockwork.Clock
+
+	ctx context.Context
+
+	cancel context.CancelFunc
+
+	// anonymizer is the anonymizer used for filtered audit events.
+	anonymizer utils.Anonymizer
+
+	// events receives batches incoming events from various Teleport components
+	events chan []*prehogapi.SubmitEventRequest
+
+	// buf stores events for batching
+	buf []*prehogapi.SubmitEventRequest
+
+	// submissionQueue queues events for submission
+	submissionQueue chan []*prehogapi.SubmitEventRequest
+
+	// submit is the func that submits batches of events to a backend
+	submit UsageSubmitFunc
+
+	// clusterName is the cluster's name, used for anonymization and as an event
+	// field.
+	clusterName types.ClusterName
+
+	// minBatchSize is the minimum batch size before a submit is triggered due
+	// to size.
+	minBatchSize int
+
+	// maxBatchSize is the maximum size of a batch that may be sent at once.
+	maxBatchSize int
+
+	// maxBatchAge is the
+	maxBatchAge time.Duration
+
+	// maxBufferSize is the maximum number of events that can be queued in the
+	// buffer.
+	maxBufferSize int
+
+	// submitDelay is the amount of delay added between all batch submission
+	// attempts.
+	submitDelay time.Duration
+
+	// receiveFunc is a callback for testing that's called when a batch has been
+	// received, but before it's been potentially enqueued, used to ensure sane
+	// sequencing in tests.
+	receiveFunc func()
+}
+
+// runSubmit starts the submission thread. It should be run as a background
+// goroutine to ensure SubmitAnonymizedUsageEvents() never blocks.
+func (r *UsageReporter) runSubmit() {
```

Comment:
I would much prefer that `ctx` is passed in as an argument to `runSubmit()` instead of being attached to `r`.